File: src/csharp/pulse/cdm/engine/SEPatientConfiguration.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

namespace Pulse.CDM
{
  public class SEPatientConfiguration
  {
    protected SEPatient patient = null;
    protected string patient_file = "";
    protected SEConditionManager condition_manager=null;
    protected string data_root_dir = "./";

    public SEPatientConfiguration()
    {
      Clear();
    }

    public void Clear()
    {
      patient_file = null;
      if(patient!=null)
        patient.Clear();
      condition_manager = null;
      data_root_dir = "./";
    }

    public bool IsValid()
    {
      if (!HasPatient() && !HasPatientFile())
        return false;
      return true;
    }

    public string GetDataRootDir()
    {
      return data_root_dir;
    }
    public void SetDataRootDir(string dir)
    {
      data_root_dir = dir;
    }

    public bool HasPatientFile()
    {
      return patient_file != null;
    }
    public string GetPatientFile()
    {
      return patient_file;
    }
    public void SetPatientFile(string f)
    {
      patient_file = f;
    }

    public bool HasPatient()
    {
      return patient != null;
    }
    public SEPatient GetPatient()
    {
      if (patient == null)
        patient = new SEPatient();
      return patient;
    }
    public void SetPatient(SEPatient p)
    {
      patient = p;
    }

    public bool HasConditions()
    {
      return condition_manager == null ? false : !condition_manager.IsEmpty();
    }
    public SEConditionManager GetConditions()
    {
      if (condition_manager == null)
        condition_manager = new SEConditionManager();
      return condition_manager;
    }
    public void InvalidateConditions()
    {
      condition_manager = null;
    }
  }
}

File: src/java/pulse/cdm/engine/SEPatientConfiguration.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package pulse.cdm.engine;

import java.util.ArrayList;
import java.util.List;

import pulse.cdm.bind.Engine.AnyConditionData;
import pulse.cdm.bind.Engine.PatientConfigurationData;
import pulse.cdm.conditions.SECondition;
import pulse.cdm.patient.SEPatient;
import pulse.cdm.substance.SESubstanceManager;

public class SEPatientConfiguration 
{
  protected SEPatient                     patient;
  protected String                        patientFile;
  protected List<SECondition>             conditions;
  
  public SEPatientConfiguration()
  {
    this.patient = null;
    this.patientFile = "";
    this.conditions = new ArrayList<>();
  }
  
  public void reset() 
  {
    patient=null;
    patientFile = "";
    conditions.clear();
  }
  
  public static void load(PatientConfigurationData src, SEPatientConfiguration dst, SESubstanceManager subMgr)
  {
    dst.reset();
    
    switch(src.getPatientTypeCase())
    {
    case PATIENT:
    {
      SEPatient p = new SEPatient();
      SEPatient.load(src.getPatient(),p);
      dst.setPatient(p);
      break;
    }
    case PATIENTFILE:
      dst.setPatientFile(src.getPatientFile());
    default: // do nothing
    }
    
    for(AnyConditionData cData : src.getConditions().getAnyConditionList())
      dst.conditions.add(SECondition.ANY2CDM(cData, subMgr));
  }
  
  public static PatientConfigurationData unload(SEPatientConfiguration src)
  {
    PatientConfigurationData.Builder dst = PatientConfigurationData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEPatientConfiguration src, PatientConfigurationData.Builder dst)
  {
    if(src.hasPatient())
      dst.setPatient(SEPatient.unload(src.patient));
    if(src.hasPatientFile())
      dst.setPatientFile(src.patientFile);
    if(src.conditions.size()>0)
    {
      for(SECondition c : src.conditions)
        dst.getConditionsBuilder().addAnyCondition(SECondition.CDM2ANY(c));    
    }
  }
  
  public boolean isValid()
  {
    if (!hasPatientFile()&&!hasPatient())
      return false;
   return true;
  }
  
  public SEPatient getPatient()
  {
    return patient;
  }
  public void setPatient(SEPatient patient)
  {
    this.patient = patient;
    this.patientFile = "";
  }
  public boolean hasPatient()
  {
    return patient==null ? false : true;
  }
  public void invalidatePatient()
  {
    patient = null;
  }

  public String getPatientFile()
  {
    return patientFile;
  }
  public void setPatientFile(String patientFile)
  {
    this.patient = null;
    this.patientFile = patientFile;
  }
  public boolean hasPatientFile()
  {
    return patientFile.isEmpty() ? false : true;
  }
  public void invalidatePatientFile()
  {
    patientFile = "";
  }
  
  public List<SECondition> getConditions() 
  {
    return conditions;
  }
}
